SerialPrograms/Source/CommonTools/OCR/OCR_NumberReader.cpp

```diff
@@ -104,52 +105,94 @@ int read_number_waterfill(
+){
+    using namespace Kernels::Waterfill;
 
-    ImageRGB32 filtered = to_blackwhite_rgb32_range(image, rgb32_min, rgb32_max, text_inside_range);
 
-//    static int c = 0;
-//    filtered.save("test-" + std::to_string(c++) + ".png");
 
-    PackedBinaryMatrix matrix = compress_rgb32_to_binary_range(filtered, 0xff000000, 0xff7f7f7f);
+    for (std::pair<uint32_t, uint32_t> filter : filters){
+
+        uint32_t rgb32_min = filter.first;
+        uint32_t rgb32_max = filter.second;
+        
+        //  Direct OCR is unreliable. Instead, we will waterfill each character
+        //  to isolate them, then OCR them individually.
+
+        ImageRGB32 filtered = to_blackwhite_rgb32_range(image, rgb32_min, rgb32_max, text_inside_range);
+
+    //    static int c = 0;
+    //    int i = 0;
+    //    filtered.save("test-" + std::to_string(c++) + ".png");
+
+        PackedBinaryMatrix matrix = compress_rgb32_to_binary_range(filtered, 0xff000000, 0xff7f7f7f);
+
+        std::map<size_t, WaterfillObject> map;
+        {
+            std::unique_ptr<WaterfillSession> session = make_WaterfillSession(matrix);
+            auto iter = session->make_iterator(20);
+            WaterfillObject object;
+            bool exceed_width_max = false;
+            while (map.size() < 16 && iter->find_next(object, true)){
+                if (object.width() > width_max){
+                    exceed_width_max = true;
+                    break;
+                }
+                map.emplace(object.min_x, std::move(object));
+            }
+            if (exceed_width_max){
+                // try the next color filter
+                continue;
+            }
+        }
 
-    std::map<size_t, WaterfillObject> map;
-    {
-        std::unique_ptr<WaterfillSession> session = make_WaterfillSession(matrix);
-        auto iter = session->make_iterator(20);
-        WaterfillObject object;
-        while (map.size() < 16 && iter->find_next(object, true)){
-            map.emplace(object.min_x, std::move(object));
+        std::string ocr_text;
+        bool empty_char = false;
+        for (const auto& item : map){
+            const WaterfillObject& object = item.second;
+            ImageRGB32 cropped = extract_box_reference(filtered, object).copy();            
+            PackedBinaryMatrix tmp(object.packed_matrix());
+            filter_by_mask(tmp, cropped, Color(0xffffffff), true);
+            ImageRGB32 padded = pad_image(cropped, cropped.width(), 0xffffffff);
+            std::string ocr = OCR::ocr_read(Language::English, padded);
+            // padded.save("test-cropped" + std::to_string(c) + "-" + std::to_string(i++) + ".png");
+            // std::cout << ocr[0] << std::endl;
+            if (!ocr.empty()){
+                ocr_text += ocr[0];
+            }else{
+                empty_char = true;
+                break;
+            }
         }
-    }
+        if (empty_char){
+            // try the next color filter
+            continue;
```

Comment:
Rather than breaking out on the first match, you might want to run all of them anyway, then do a majority vote on the results (after throwing out obvious incorrect reads like empty string or letters). TeraCardReader does this with the star count: https://github.com/PokemonAutomation/Arduino-Source/blob/ba9d5bdc7254387bc601dfde8675e20250f6b9e7/SerialPrograms/Source/PokemonSV/Inference/Tera/PokemonSV_TeraCardDetector.cpp#L132

The reason is that on the borderline cases may return neither an empty string nor a valid result. You'll want to result when the "signal is strongest".